packages/lit-html/src/lit-html.ts

```diff
@@ -711,6 +715,48 @@ class Template {
   }
 }
 
+function resolveDirective(
+  part: NodePart | AttributePart,
+  value: unknown,
+  directiveParent: DirectiveParent = part,
+  attributeIndex?: number
+): unknown {
+  let currentDirective =
+    attributeIndex !== undefined
+      ? (directiveParent as AttributePart)._directives?.[attributeIndex]
+      : (directiveParent as NodePart | Directive)._directive;
+  const nextDirectiveConstructor = isPrimitive(value)
+    ? undefined
+    : (value as DirectiveResult)?._$litDirective$;
+  if (currentDirective?.constructor !== nextDirectiveConstructor) {
+    if (nextDirectiveConstructor !== undefined) {
+      currentDirective = new nextDirectiveConstructor(
+        part as PartInfo,
+        attributeIndex
+      );
+      currentDirective._part = part;
+      currentDirective._attributeIndex = attributeIndex;
```

Comment:
This is just a choice. We can do this in the constructor since we're passing them already, but it requires the user to super them through, which is sorta annoying (directives may need to set up their own state in the ctor, but normally they only need these arguments in limited cases).

Reply:
I'm fine with that, will go ahead and update ll the directives to do this.